testsuite/tests/lib-random/parallel.ml

```diff
@@ -0,0 +1,27 @@
+(* TEST *)
+let () = Random.init 42
+
+let f () =
+  let a = Random.int 100 in
+  let b = Random.int 100 in
+  let c = Random.int 100 in
+  (a, b, c)
+
+(* We intentionally spawn many more domains than hardware threads, to
```

Comment:
I thought the maximal number of domains is 128.  In this case it's risky to spawn 1000.  I guess it works because each domain exits almost immediately, so domains die faster than they are created.  But in this case the OS scheduler doesn't get much chance to interleave executions, does it?

Bottom line: I'd rather have a test with a small number of domains, and each domain waits a variable -- maybe even random! -- amount of time before drawing random numbers.

Reply:
> I'd rather have a test with a small number of domains, and each domain waits a variable -- maybe even random! -- amount of time before drawing random numbers.

Took the liberty to push the revised test on this PR's branch.

Reply:
I'm surprised to learn that domains are limited to 128, and that my test was not reaching that limit. I'll double-check. (128 seems a bit low, I could buy AMD Threadripper machines with that many threads today.)

Reply:
> The new implementation does all the computation on the parent

What?  One of us is missing something.

Reply:
I meant that `split_from_parent` executes on the parent domain: for inherited/derived keys there is no part of the DLS value initialization protocol that is performed by the child, and could end up interleaved with other domains sharing the same parent.

Reply:
OK, I got your point.  We can also delete this test if it's no longer relevant.

> 'm not sure what failure modes we can observe with the revised test;

It would fail if
1) a single PRNG was shared between all domains (with a mutex for protection)
2) each domain started with the same PRNG seed
3) each domain was running self-initialization when started.

Reply:
> I'm surprised to learn that domains are limited to 128

We needed max domains to be a small number for fast read barrier checks. See the last paragraph in section 4.3.2 in https://arxiv.org/pdf/2004.11663.pdf. While we no longer use read barriers, we still have the virtual memory layout as it turns out to be good for minimizing memory hierarchy overheads. See https://github.com/ocaml-multicore/ocaml-multicore/pull/508 for the experiments. The max domains value affects the size of the virtual address space reserved. 

I have a proposal for removing this restriction here: https://github.com/ocaml-multicore/ocaml-multicore/issues/795. For example, you could have max domains to be 256 with a max minor heap size of 2M, and the program will only reserve 512M of virtual memory space.  Currently, we reserve 256G.

Reply:
I integrated @xavierleroy's proposed test in the commit history. The test doesn't allow to detect races if several child are initialized at the same time, but it does detect that seeding is performed (that the random stream does not depend on the scheduling order of random draws).

Reply:
The commit message for af8a9065f5dc6c8c77998107c08c02b5aafed8bc still describes the previous, 1000-domain version of the test.